### PR TITLE
Use built-in JSON parser for PXF fragments

### PR DESCRIPTION
**Pull request by danielgustafsson**
Instead of relying on a external library, and thus yet another dependency, use the built-in JSON
parser in the backend for the PXF fragments parsing. Since this replaces the current implementation with an event-based callback parser, the code is more complicated, but dogfooding the parser that we want extension writers to use is a good thing.

There are still some rough edges, but it's a first step. I also don't really know how to test it so publishing here in hope of some assistance and more eyes looking on it.

If the fragment doesn't have any replicas, the list entry is dropped but not freed currently. This leaks the list entry but will be cleaned up by the memory manager when the context closes (this can of course be done explicitly here as well should this scenario be considered common).

### CODE REVIEW

**Comment by hsyuan** [comment]
Nice!

**Comment by shivzone** [comment]
gpAux/extensions/pxf/test/Makefile still links to json-c library

**Comment by shivzone** [comment]
@danielgustafsson I've attached a patch which fixes the issues related to unit tests. The SAX based parser works great. If you can merge this patch to your branch, we can create a pipeline to validate the regression tests as well.
[patch.txt](https://github.com/greenplum-db/gpdb/files/1283079/patch.txt)

File: gpAux/extensions/pxf/src/pxffragment.c [hunk 1]
```diff
@@ -176,76 +180,206 @@ rest_request(GPHDUri *hadoop_uri, ClientContext *client_context, char *rest_msg)
...
  */
-static List *
-parse_get_fragments_response(List *fragments, StringInfo rest_buf)
+
+typedef enum pxf_fragment_object
+{
+	PXF_START,
+	PXF_INDEX,
+	PXF_USERDATA,
+	PXF_PROFILE_,
```
Comment (hlinnaka) [inline gpAux/extensions/pxf/src/pxffragment.c]: Is the underscore here by accident?

**Comment by danielgustafsson** [comment]
> On 07 Sep 2017, at 07:49, Heikki Linnakangas <notifications@github.com> wrote:
> 
> @hlinnaka approved this pull request.
> 
> Looks good at a quick glance.

Thanks!

> In gpAux/extensions/pxf/src/pxffragment.c:
> 
> > @@ -176,76 +180,206 @@ rest_request(GPHDUri *hadoop_uri, ClientContext *client_context, char *rest_msg)
>   * Response (left as a single line purposefully):
>   * {"PXFFragments":[{"index":0,"userData":null,"sourceName":"demo/text2.csv","metadata":"rO0ABXcQAAAAAAAAAAAAAAAAAAAABXVyABNbTGphdmEubGFuZy5TdHJpbmc7rdJW5+kde0cCAAB4cAAAAAN0ABxhZXZjZWZlcm5hczdtYnAuY29ycC5lbWMuY29tdAAcYWV2Y2VmZXJuYXM3bWJwLmNvcnAuZW1jLmNvbXQAHGFldmNlZmVybmFzN21icC5jb3JwLmVtYy5jb20=","replicas":["10.207.4.23","10.207.4.23","10.207.4.23"]},{"index":0,"userData":null,"sourceName":"demo/text_csv.csv","metadata":"rO0ABXcQAAAAAAAAAAAAAAAAAAAABnVyABNbTGphdmEubGFuZy5TdHJpbmc7rdJW5+kde0cCAAB4cAAAAAN0ABxhZXZjZWZlcm5hczdtYnAuY29ycC5lbWMuY29tdAAcYWV2Y2VmZXJuYXM3bWJwLmNvcnAuZW1jLmNvbXQAHGFldmNlZmVybmFzN21icC5jb3JwLmVtYy5jb20=","replicas":["10.207.4.23","10.207.4.23","10.207.4.23"]}]}
>   */
> -static List *
> -parse_get_fragments_response(List *fragments, StringInfo rest_buf)
> +
> +typedef enum pxf_fragment_object
> +{
> +	PXF_START,
> +	PXF_INDEX,
> +	PXF_USERDATA,
> +	PXF_PROFILE_,
> 
> Is the underscore here by accident?

Yes and no. PXF_PROFILE was already defined elsewhere, but I intended to replace the temporary PXF_PROFILE_ with something better once I had figured out a good name. But that slipped my mind it seems.

**Comment by hlinnaka** [comment]
On 09/07/2017 08:58 AM, Daniel Gustafsson wrote:
>> On 07 Sep 2017, at 07:49, Heikki Linnakangas <notifications@github.com> wrote:
>>
>> @hlinnaka approved this pull request.
>>
>> Looks good at a quick glance.
> 
> Thanks!
> 
>> In gpAux/extensions/pxf/src/pxffragment.c:
>>
>>> @@ -176,76 +180,206 @@ rest_request(GPHDUri *hadoop_uri, ClientContext *client_context, char *rest_msg)
>>    * Response (left as a single line purposefully):
>>    * {"PXFFragments":[{"index":0,"userData":null,"sourceName":"demo/text2.csv","metadata":"rO0ABXcQAAAAAAAAAAAAAAAAAAAABXVyABNbTGphdmEubGFuZy5TdHJpbmc7rdJW5+kde0cCAAB4cAAAAAN0ABxhZXZjZWZlcm5hczdtYnAuY29ycC5lbWMuY29tdAAcYWV2Y2VmZXJuYXM3bWJwLmNvcnAuZW1jLmNvbXQAHGFldmNlZmVybmFzN21icC5jb3JwLmVtYy5jb20=","replicas":["10.207.4.23","10.207.4.23","10.207.4.23"]},{"index":0,"userData":null,"sourceName":"demo/text_csv.csv","metadata":"rO0ABXcQAAAAAAAAAAAAAAAAAAAABnVyABNbTGphdmEubGFuZy5TdHJpbmc7rdJW5+kde0cCAAB4cAAAAAN0ABxhZXZjZWZlcm5hczdtYnAuY29ycC5lbWMuY29tdAAcYWV2Y2VmZXJuYXM3bWJwLmNvcnAuZW1jLmNvbXQAHGFldmNlZmVybmFzN21icC5jb3JwLmVtYy5jb20=","replicas":["10.207.4.23","10.207.4.23","10.207.4.23"]}]}
>>    */
>> -static List *
>> -parse_get_fragments_response(List *fragments, StringInfo rest_buf)
>> +
>> +typedef enum pxf_fragment_object
>> +{
>> +	PXF_START,
>> +	PXF_INDEX,
>> +	PXF_USERDATA,
>> +	PXF_PROFILE_,
>>
>> Is the underscore here by accident?
> 
> Yes and no. PXF_PROFILE was already defined elsewhere, but I intended to replace the temporary PXF_PROFILE_ with something better once I had figured out a good name. But that slipped my mind it seems.

Oh, ok. Perhaps a different prefix would be better here then, to avoid 
confusion with other PXF_* things.

- Heikki

**Comment by danielgustafsson** [comment]
> On 07 Sep 2017, at 08:14, Heikki Linnakangas <notifications@github.com> wrote:
> 
> On 09/07/2017 08:58 AM, Daniel Gustafsson wrote:
> >> On 07 Sep 2017, at 07:49, Heikki Linnakangas <notifications@github.com> wrote:
> >>
> >> @hlinnaka approved this pull request.
> >>
> >> Looks good at a quick glance.
> > 
> > Thanks!
> > 
> >> In gpAux/extensions/pxf/src/pxffragment.c:
> >>
> >>> @@ -176,76 +180,206 @@ rest_request(GPHDUri *hadoop_uri, ClientContext *client_context, char *rest_msg)
> >> * Response (left as a single line purposefully):
> >> * {"PXFFragments":[{"index":0,"userData":null,"sourceName":"demo/text2.csv","metadata":"rO0ABXcQAAAAAAAAAAAAAAAAAAAABXVyABNbTGphdmEubGFuZy5TdHJpbmc7rdJW5+kde0cCAAB4cAAAAAN0ABxhZXZjZWZlcm5hczdtYnAuY29ycC5lbWMuY29tdAAcYWV2Y2VmZXJuYXM3bWJwLmNvcnAuZW1jLmNvbXQAHGFldmNlZmVybmFzN21icC5jb3JwLmVtYy5jb20=","replicas":["10.207.4.23","10.207.4.23","10.207.4.23"]},{"index":0,"userData":null,"sourceName":"demo/text_csv.csv","metadata":"rO0ABXcQAAAAAAAAAAAAAAAAAAAABnVyABNbTGphdmEubGFuZy5TdHJpbmc7rdJW5+kde0cCAAB4cAAAAAN0ABxhZXZjZWZlcm5hczdtYnAuY29ycC5lbWMuY29tdAAcYWV2Y2VmZXJuYXM3bWJwLmNvcnAuZW1jLmNvbXQAHGFldmNlZmVybmFzN21icC5jb3JwLmVtYy5jb20=","replicas":["10.207.4.23","10.207.4.23","10.207.4.23"]}]}
> >> */
> >> -static List *
> >> -parse_get_fragments_response(List *fragments, StringInfo rest_buf)
> >> +
> >> +typedef enum pxf_fragment_object
> >> +{
> >> +	PXF_START,
> >> +	PXF_INDEX,
> >> +	PXF_USERDATA,
> >> +	PXF_PROFILE_,
> >>
> >> Is the underscore here by accident?
> > 
> > Yes and no. PXF_PROFILE was already defined elsewhere, but I intended to replace the temporary PXF_PROFILE_ with something better once I had figured out a good name. But that slipped my mind it seems.
> 
> Oh, ok. Perhaps a different prefix would be better here then, to avoid 
> confusion with other PXF_* things.

I was thinking on using PXF_PARSER_* instead for all these enums, to make it clear that these are states for the parsing.

**Comment by danielgustafsson** [comment]
> On 07 Sep 2017, at 02:21, Shivram Mani <notifications@github.com> wrote:
> 
> @danielgustafsson I've attached a patch which fixes the issues related to unit tests. The SAX based parser works great. If you can merge this patch to your branch, we can create a pipeline to validate the regression tests as well.
> patch.txt

The patch has been merged, thanks for the help! Current PR runs all the unittests for me, please have a test against the regression tests.

File: gpAux/extensions/pxf/test/pxffragment_test.c [hunk 2]
```diff
@@ -304,7 +304,7 @@ test_parse_get_fragments_response_nulluserdata(void **state)
 	StringInfoData frag_json;
 
 	initStringInfo(&frag_json);
-	appendStringInfo(&frag_json, "{\"PXFFragments\":[{\"index\":1,\"userData\":null,\"sourceName\":\"demo/text2.csv\",\"metadata\":\"metadatavalue1\",\"replicas\":[\"localhost\",\"localhost\",\"localhost\"],\"profile\":\"DemoProfile\"}]}");
+	appendStringInfo(&frag_json, "{\"PXFFragments\":[{\"index\":1,\"userData\":{},\"sourceName\":\"demo/text2.csv\",\"metadata\":\"metadatavalue1\",\"replicas\":[\"localhost\",\"localhost\",\"localhost\"],\"profile\":\"DemoProfile\"}]}");
```
Comment (denalex) [inline gpAux/extensions/pxf/test/pxffragment_test.c]: I don't think this is the right change to unit tests. The problem we faced that the new parser would parse JSON null into a string "null", which is incorrect. Changing the test input to get around this is a hack, instead we should get to the bottom of the issue.

**Comment by danielgustafsson** [comment]
> On 07 Sep 2017, at 18:57, Alexander Denissov <notifications@github.com> wrote:
> 
> @denalex commented on this pull request.
> 
> In gpAux/extensions/pxf/test/pxffragment_test.c:
> 
> > @@ -304,7 +304,7 @@ test_parse_get_fragments_response_nulluserdata(void **state)
>  	StringInfoData frag_json;
>  
>  	initStringInfo(&frag_json);
> -	appendStringInfo(&frag_json, "{\"PXFFragments\":[{\"index\":1,\"userData\":null,\"sourceName\":\"demo/text2.csv\",\"metadata\":\"metadatavalue1\",\"replicas\":[\"localhost\",\"localhost\",\"localhost\"],\"profile\":\"DemoProfile\"}]}");
> +	appendStringInfo(&frag_json, "{\"PXFFragments\":[{\"index\":1,\"userData\":{},\"sourceName\":\"demo/text2.csv\",\"metadata\":\"metadatavalue1\",\"replicas\":[\"localhost\",\"localhost\",\"localhost\"],\"profile\":\"DemoProfile\"}]}");
> 
> I don't think this is the right change to unit tests. The problem we faced that the new parser would parse JSON null into a string "null", which is incorrect. Changing the test input to get around this is a hack, instead we should get to the bottom of the issue.

Fair enough, I only applied the diff provided. What should happen on JSON null in the fragment? Should the string in the FragmentData struct be a NULL pointer or should be a literal null? Either way I can easily fix the code, as long as I know what you want.

**Comment by lavjain** [comment]
@danielgustafsson I worked with @denalex to fix the null handling. Please take a look at
https://github.com/danielgustafsson/gpdb/pull/7 and feel free to merge this PR.

**Comment by lavjain** [comment]
@danielgustafsson Could you please look at this PR and merge it soon?

**Comment by danielgustafsson** [comment]
> On 12 Sep 2017, at 19:58, Lav Jain <notifications@github.com> wrote:
> 
> @danielgustafsson Could you please look at this PR and merge it soon?

Sorry for the latency, I’ve been travelling and haven’t kept up. I’ve merged your patch with some massaging and also enabled building PXF on Travis since that should be supported now. If this builds green I think we are ready to close this out and merge to master.

**Comment by lavjain** [comment]
Please feel free to merge

**Comment by danielgustafsson** [comment]
This fails the PR pipeline, but I can't find the test that actually runs?

**Comment by lavjain** [comment]
The pipeline was green with my branch that was based off yours. I did not see any updates on your branch yet.

**Comment by danielgustafsson** [comment]
> On 13 Sep 2017, at 17:40, Lav Jain <notifications@github.com> wrote:
> 
> The pipeline was green with my branch that was based off yours. I did not see any updates on your branch yet.

The only changes introduced when applying your patch which werent cosmetical, was to error out on unknown PARSE states:

https://github.com/greenplum-db/gpdb/pull/3185/files#diff-7ed50ef56448d6b76476e6e998551da5R277

The cause of the failure in the PR pipeline isn’t printed to the pipeline output so one is left to guesswork, but the change above should not cause (if it did we have another bug).

**Comment by lavjain** [comment]
I believe it is safe to merge. Please merge to master as well as 5X_STABLE

**Comment by lavjain** [comment]
Cherry-picked commit to 5X_STABLE as well.

**Pull request merged**